File: iec61131lang/src/test/java/edu/kit/iti/formal/automation/st/InValidExpressionTest.java
package edu.kit.iti.formal.automation.st;

/*-
 * #%L
 * iec61131lang
 * %%
 * Copyright (C) 2016 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import edu.kit.iti.formal.automation.TestUtils;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

/**
 * @author Alexander Weigl
 * @version 1 (02.08.16)
 */
@RunWith(Parameterized.class)
public class InValidExpressionTest {
    @Parameterized.Parameters(name = "{0}")
    public static Iterable<Object[]> setup() throws IOException {
        return TestUtils.loadLines("/edu/kit/iti/formal/automation/st/expressions/invalid.txt");
    }

    @Parameterized.Parameter
    public String invalid = "";

    @Test
    public void testInValidExpression() {
        assertFalse(ValidExpressionTest.test(invalid));
    }
}

File: lang/src/test/kotlin/edu/kit/iti/formal/automation/InValidExpressionTest.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 *
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.automation

import TestUtils
import org.antlr.v4.runtime.CharStreams
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource

/**
 * @author Alexander Weigl
 * @version 1 (02.08.16)
 */
object InValidExpressionTest {

    @ParameterizedTest
    @MethodSource("getExpressions")
    fun testInValidExpression(invalid: String) {
        val parser = IEC61131Facade.getParser(CharStreams.fromString(invalid))
        var error = try {
            val ctx = parser.expression()
            println(ctx.toString(parser))
            println(IEC61131Facade.print(IEC61131Facade.expr(invalid)))
            println(ctx.stop.tokenSource.nextToken())
            ctx.stop.tokenSource.nextToken().text != "EOF"
            // error = false;
        } catch (e: Exception) {
            true
        }

        error = error || parser.numberOfSyntaxErrors > 0
        println(parser.numberOfSyntaxErrors)
        assertTrue(error)
    }

    @JvmStatic
    fun getExpressions() = TestUtils.loadLines("/edu/kit/iti/formal/automation/st/expressions/invalid.txt")
}
